### PR TITLE
HubSpot CRM

### PR DESCRIPTION
**Pull request by troystaylor**
---
When submitting a connector, please check the following conditions for your PR.

- [X] I attest that the connector works and I verified by deploying and testing all the operations.
- [X] I validated the swagger file, `apiDefinition.swagger.json`, by running `paconn validate` command.
- [X] If this is a certified connector, I confirm that `apiProperties.json` has a valid brand color and doesn't use an invalid brand color, `#007ee5` or `#ffffff`. If this is an independent publisher connector, I confirm that I am not submitting a logo.

### CODE REVIEW

File: independent-publisher-connectors/HubSpot CRM/apiDefinition.swagger.json [hunk 1]
```diff
@@ -0,0 +1,4553 @@
+{
+  "swagger": "2.0",
+  "info": {
+    "title": "Default title",
```
Comment (sriyen-msft) [inline independent-publisher-connectors/HubSpot CRM/apiDefinition.swagger.json:4]: @troystaylor Please provide a valid title

File: independent-publisher-connectors/HubSpot CRM/apiDefinition.swagger.json [hunk 2]
```diff
@@ -0,0 +1,4553 @@
...
+                      "type": "string",
+                      "description": "Marital status."
+                    },
+                    "message": {
+                      "title": "Message",
+                      "type": "string",
+                      "description": "Message."
+                    },
+                    "military_status": {
+                      "title": "Military statusMilitary status",
```
Comment (sriyen-msft) [inline independent-publisher-connectors/HubSpot CRM/apiDefinition.swagger.json:1780]: @troystaylor Typo? All the typos in swagger need to be fixed.

**Review by sriyen-msft — CHANGES_REQUESTED** [review]

File: independent-publisher-connectors/HubSpot CRM/apiProperties.json [hunk 3]
```diff
@@ -0,0 +1,75 @@
+{
```
Comment (sriyen-msft) [inline independent-publisher-connectors/HubSpot CRM/apiProperties.json:1]: @troystaylor Did you tested this connector in your dev/local enviornment, If yes, didn't you get issues ?

**Review by sriyen-msft — CHANGES_REQUESTED** [review]

File: independent-publisher-connectors/HubSpot CRM/apiDefinition.swagger.json (hunk 1 above)
Comment (troystaylor) [reply independent-publisher-connectors/HubSpot CRM/apiDefinition.swagger.json:4]: The title I see is:
"HubSpot CRM is a system for managing customer relationships and storing data. The CRM custom connector provides access to CRM objects or groups of objects, which are represented as a map of property names to values."

File: independent-publisher-connectors/HubSpot CRM/apiDefinition.swagger.json (hunk 2 above)
Comment (troystaylor) [reply independent-publisher-connectors/HubSpot CRM/apiDefinition.swagger.json]: Fixed

File: independent-publisher-connectors/HubSpot CRM/apiProperties.json (hunk 3 above)
Comment (troystaylor) [reply independent-publisher-connectors/HubSpot CRM/apiProperties.json:1]: Copied properties from CMS connector, fixed

Comment (troystaylor) [reply independent-publisher-connectors/HubSpot CRM/apiProperties.json:1]: Updated

File: independent-publisher-connectors/HubSpot CRM/apiDefinition.swagger.json (hunk 1 above)
Comment (sriyen-msft) [reply independent-publisher-connectors/HubSpot CRM/apiDefinition.swagger.json:4]: > The title I see is:
> "HubSpot CRM is a system for managing customer relationships and storing data. The CRM custom connector provides access to CRM objects or groups of objects, which are represented as a map of property names to values."

This is the description value.

The title value need to be changed.

line number 4:  `"title": "Default title"`

**Review by sriyen-msft — CHANGES_REQUESTED** [review]

**Comment by troystaylor** [comment]
Fixed title

**Review by mamurshe — CHANGES_REQUESTED** [review]
When I try to create a custom connector using these asset with paconn CLI I get an error. Can you please update the apiProperties.json using `paconn download` command? 

```
independent-publisher-connectors\HubSpot CRM>paconn create -d apiDefinition.swagger.json -p apiProperties.json -e XXXXXXXX-XXXX-XXXX-XXXX-XXXXXXXXXXXX
'properties'
Traceback (most recent call last):
  File "c:\python\lib\site-packages\knack\cli.py", line 206, in invoke
    cmd_result = self.invocation.execute(args)
  File "c:\python\lib\site-packages\knack\invocation.py", line 188, in execute
    cmd_result = parsed_args.func(params)
  File "c:\python\lib\site-packages\knack\commands.py", line 131, in __call__
    return self.handler(*args, **kwargs)
  File "c:\python\lib\site-packages\knack\commands.py", line 238, in _command_handler
    result = op(client, **command_args) if client else op(**command_args)
  File "c:\python\lib\site-packages\paconn\commands\create.py", line 47, in create
    connector_id = upsert(
  File "c:\python\lib\site-packages\paconn\operations\upsert.py", line 82, in upsert
    properties = property_definition[_PROPERTIES]
KeyError: 'properties'
```

**Review by mamurshe — CHANGES_REQUESTED** [review]
"redirectUrl": ["https://global.consent.azure-apim.net/redirect", "https://global-test.consent.azure-apim.net/redirect"],

to
"redirectUrl": "https://global.consent.azure-apim.net/redirect",

**Review by mamurshe — CHANGES_REQUESTED** [review]
Can you please add `connectionParameters` section as shown below? It's needed for deployment.

```
  "properties": {
    "connectionParameters": {
      "api_key": {
        "type": "securestring",
        "uiDefinition": {
          "constraints": {
            "clearText": false,
            "required": "true",
            "tabIndex": 3
          },
          "schema": {
            "description": "Enter your API Key for HubSpot",
            "type": "securestring"
          },
          "displayName": "API Key generated in HubSpot"
        }
      }
    },
    "connectionParameterSets": {
```

**Comment by troystaylor** [comment]
> Can you please add `connectionParameters` section as shown below? It's needed for deployment.
> 
> ```
>   "properties": {
>     "connectionParameters": {
>       "api_key": {
>         "type": "securestring",
>         "uiDefinition": {
>           "constraints": {
>             "clearText": false,
>             "required": "true",
>             "tabIndex": 3
>           },
>           "schema": {
>             "description": "Enter your API Key for HubSpot",
>             "type": "securestring"
>           },
>           "displayName": "API Key generated in HubSpot"
>         }
>       }
>     },
>     "connectionParameterSets": {
> ```

I don't understand why OAuth would be put in a connectionParameterSet and not the API key

File: independent-publisher-connectors/HubSpot CRM/apiProperties.json [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+            "description": "Log in using HubSpot app client ID and secret."
+          },
+          "parameters": {
+            "token": {
+              "type": "oauthSetting",
+              "oAuthSettings": {
+                "identityProvider": "oauth2",
+                "clientId": "[Your_Value]",
+                "scopes": [
+                  "contacts oauth"
```
Comment (mamurshe) [inline independent-publisher-connectors/HubSpot CRM/apiProperties.json:22]: The scopes here and in the swagger file don't match. Can you please update the correct scopes? ApiProperties is the one used in down stream application. Please update all HubSpot connector with similar changes.

File: independent-publisher-connectors/HubSpot CRM/apiProperties.json [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+          }
+        },
+        {
+          "name": "api-auth",
+          "uiDefinition": {
+            "displayName": "Use HubSpot API Key",
+            "description": "Log in using a HubSpot API Key."
+          },
+          "parameters": {
+            "api_key": {
```
Comment (mamurshe) [inline independent-publisher-connectors/HubSpot CRM/apiProperties.json:51]: According to the HubSpot documentation:
https://legacydocs.hubspot.com/docs/methods/auth/oauth-overview
```
API keys are great for rapid prototyping, but for security and commercial use, all integrations designed for use by multiple HubSpot customers should use OAuth—this is a requirement to be listed in our App Marketplace.
```
Also the api key is named as `hapikey` and provided as part of the query parameter. Please either remove api key auth or update this connection method. Please test your changes before updating. Please make similar change for all HubSpot connectors.

File: independent-publisher-connectors/HubSpot CRM/apiDefinition.swagger.json [hunk 6]
```diff
@@ -0,0 +1,4553 @@
...
+  },
+  "definitions": {},
+  "parameters": {},
+  "responses": {},
+  "securityDefinitions": {
+    "oauth2_auth": {
+      "type": "oauth2",
+      "flow": "accessCode",
+      "authorizationUrl": "https://app.hubspot.com/oauth/authorize",
+      "tokenUrl": "https://api.hubapi.com/oauth/v1/token",
```
Comment (mamurshe) [inline independent-publisher-connectors/HubSpot CRM/apiDefinition.swagger.json:4525]: Add refresh url.

File: independent-publisher-connectors/HubSpot CRM/apiDefinition.swagger.json [hunk 7]
```diff
@@ -0,0 +1,4553 @@
...
+  },
+  "host": "api.hubapi.com",
+  "basePath": "/",
+  "schemes": [
+    "https"
+  ],
+  "consumes": [],
+  "produces": [],
+  "paths": {
+    "/crm/v3/objects/companies": {
```
Comment (mamurshe) [inline independent-publisher-connectors/HubSpot CRM/apiDefinition.swagger.json:21]: /crm/v3/objects/ should be base path.

**Review by mamurshe — CHANGES_REQUESTED** [review]

File: independent-publisher-connectors/HubSpot CRM/apiDefinition.swagger.json (hunk 6 above)
Comment (troystaylor) [reply independent-publisher-connectors/HubSpot CRM/apiDefinition.swagger.json:4525]: Fixed

File: independent-publisher-connectors/HubSpot CRM/apiDefinition.swagger.json (hunk 7 above)
Comment (troystaylor) [reply independent-publisher-connectors/HubSpot CRM/apiDefinition.swagger.json:21]: By leaving this as base path "/", additional action based on v1 and beta actions can be developed. No change

File: independent-publisher-connectors/HubSpot CRM/apiProperties.json (hunk 5 above)
Comment (troystaylor) [reply independent-publisher-connectors/HubSpot CRM/apiProperties.json]: Fixed all properties to hapikey. HubSpot documentation is only providing security guidance - some companies may be comfortable with using an API key for their security. I don't expect any commercial users of these custom connectors to list them on the marketplace. Additional direction from Natalie requested as multi-auth was required for these connectors.

File: independent-publisher-connectors/HubSpot CRM/apiProperties.json [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+            "description": "Log in using HubSpot app client ID and secret."
+          },
+          "parameters": {
+            "token": {
+              "type": "oauthSetting",
+              "oAuthSettings": {
+                "identityProvider": "oauth2",
+                "clientId": "[Your_Value]",
+                "scopes": [
+                  "crm.objects.marketing_events.read contacts content reports actions timeline oauth forms files crm.objects.marketing_events.write crm.schemas.custom.read crm.objects.custom.read crm.objects.feedback_submissions.read crm.objects.custom.write": "crm.objects.marketing_events.read contacts content reports actions timeline oauth forms files crm.objects.marketing_events.write crm.schemas.custom.read crm.objects.custom.read crm.objects.feedback_submissions.read crm.objects.custom.write"
```
Comment (mamurshe) [inline independent-publisher-connectors/HubSpot CRM/apiProperties.json:22]: Please correct col 254. The scopes are repeated.

**Review by mamurshe — CHANGES_REQUESTED** [review]

File: independent-publisher-connectors/HubSpot CRM/apiProperties.json (hunk 8 above)
Comment (troystaylor) [reply independent-publisher-connectors/HubSpot CRM/apiProperties.json]: Fixed

**Comment by sriyen-msft** [comment]
[[certify-connector]]

File: independent-publisher-connectors/HubSpot CRM/apiProperties.json (hunk 3 above)
Comment (mamurshe) [inline independent-publisher-connectors/HubSpot CRM/apiProperties.json:1]: Please remove apikey auth.
```
{
  "properties": {
    "connectionParameters": {
      "token": {
        "type": "oauthSetting",
        "oAuthSettings": {
          "identityProvider": "oauth2",
          "clientId": "[Your_Value]",
          "scopes": [
            "crm.objects.marketing_events.read contacts content reports actions timeline oauth forms files crm.objects.marketing_events.write crm.schemas.custom.read crm.objects.custom.read crm.objects.feedback_submissions.read crm.objects.custom.write"
          ],
          "redirectMode": "Global",
          "redirectUrl": "https://global.consent.azure-apim.net/redirect",
          "properties": {
            "IsFirstParty": "False"
          },
          "customParameters": {
            "authorizationUrl": {
              "value": "https://app.hubspot.com/oauth/authorize"
            },
            "tokenUrl": {
              "value": "https://api.hubapi.com/oauth/v1/token"
            },
            "refreshUrl": {
              "value": "https://api.hubapi.com/oauth/v1/token"
            }
          }
        }
      }
    },
    "iconBrandColor": "#da3b01",
    "capabilities": [],
    "publisher": "Hitachi Solutions",
    "stackOwner": "HubSpot"
  }
}
```

File: independent-publisher-connectors/HubSpot CRM/apiDefinition.swagger.json [hunk 9]
```diff
@@ -0,0 +1,4554 @@
...
+            }
+          }
+        ]
+      }
+    }
+  },
+  "definitions": {},
+  "parameters": {},
+  "responses": {},
+  "securityDefinitions": {
```
Comment (mamurshe) [inline independent-publisher-connectors/HubSpot CRM/apiDefinition.swagger.json:4520]: Please remove `securityDefinitions` and `security` sections.

**Review by mamurshe — CHANGES_REQUESTED** [review]

**Pull request merged**